taichi/backends/opengl/opengl_api.cpp

```diff
@@ -285,26 +287,106 @@ struct GLBuffer : GLSSBO {
   }
 };
 
+struct GLBufferAllocator {
+ private:
+  const size_t max_free_resident_size = 64 << 20;
+
+  std::list<std::unique_ptr<GLBuffer>> buffers_storage;
+
+  std::map<std::pair<GLBufId, size_t>, GLBuffer *> buffers_mapping;
+  std::map<std::pair<GLBufId, size_t>, GLBuffer *> free_mapping;
+
+ public:
+  GLBufferAllocator() {
+  }
+
+  void new_frame() {
+    size_t free_resident_size = 0;
+    for (auto &pair : free_mapping) {
+      free_resident_size += pair.first.second;
+
+      if (free_resident_size > max_free_resident_size) {
+        GLBuffer *buf = pair.second;
+
+        auto buffer_iter =
+            std::find_if(buffers_storage.begin(), buffers_storage.end(),
+                         [buf](const auto &mo) { return mo.get() == buf; });
+
+        buffers_storage.erase(buffer_iter);
+        free_mapping.erase(pair.first);
+      }
+    }
+  }
+
+  GLBuffer *alloc_buffer(GLBufId index, void *base, size_t size) {
+    GLBuffer *buffer;
+    auto buffer_iter = free_mapping.find(std::pair(index, size));
+    if (buffer_iter == free_mapping.end()) {
+      // This buffer does not exist / can not be reused
+      buffers_storage.push_back(std::make_unique<GLBuffer>(index, base, size));
+      buffer = buffers_storage.back().get();
+      buffers_mapping[std::pair(index, size)] = buffer;
+
+      // TI_WARN("New buffer {}, {}, {}", int(index), size, (void*)buffer);
+    } else {
+      // Reuse
+      buffer = buffer_iter->second;
+      free_mapping.erase(buffer_iter);
+
+      buffer->rebind(base, size);
+      buffer->bind_index(int(index));
+
```

Comment:
Update `buffers_mapping` as well?

Reply:
Since `size` and `index` are kept constant during reuse, the entry within `buffers_mapping` should be constant as well

Reply:
Then we should clean up `buffers_mapping` in `new_frame()`.

Reply:
Ah right